File: android/src/main/kotlin/br/com/felnanuke2/media_cast_dlna/MediaCastDlnaPlugin.kt
package br.com.felnanuke2.media_cast_dlna

import DeviceUdn
import DiscoveryOptions
import DlnaDevice
import DlnaService
import MediaCastDlnaApi
import MediaMetadata
import PlaybackInfo
import SubtitleTrack
import TransportState
import VolumeInfo
import VolumeLevel
import MuteOperation
import MuteState
import TimePosition
import TimeDuration
import Url
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.ComponentName
import android.os.IBinder
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import kotlinx.coroutines.TimeoutCancellationException
import org.jupnp.android.AndroidUpnpService
import org.jupnp.android.AndroidUpnpServiceImpl
import br.com.felnanuke2.media_cast_dlna.core.DidlMetadataConverter
import br.com.felnanuke2.media_cast_dlna.core.DefaultDidlMetadataConverter
import br.com.felnanuke2.media_cast_dlna.core.MediaControlManager
import br.com.felnanuke2.media_cast_dlna.core.DeviceDiscoveryManager
import br.com.felnanuke2.media_cast_dlna.core.VolumeManager

/** MediaCastDlnaPlugin - Refactored to act as a simple Facade with coroutines support */
class MediaCastDlnaPlugin : FlutterPlugin, MediaCastDlnaApi {
    private var context: Context? = null
    private var upnpService: AndroidUpnpService? = null
    private var isServiceBound = false
    private var upnpRegistryListener: UpnpRegistryListener = UpnpRegistryListener()

    // Coroutine scope for managing async operations
    private val pluginScope =
        CoroutineScope(SupervisorJob() + Dispatchers.IO) // Changed to IO dispatcher

    // Service connection for UPnP service
    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            upnpService = service as AndroidUpnpService
            upnpService?.get()?.startup()
            upnpService?.registry?.addListener(upnpRegistryListener)
            isServiceBound = true
        }

        override fun onServiceDisconnected(className: ComponentName) {
            upnpService = null
            isServiceBound = false
        }
    }

    private val didlMetadataConverter: DidlMetadataConverter = DefaultDidlMetadataConverter()
    private val mediaControlManager: MediaControlManager by lazy {
        MediaControlManager(upnpService)
    }
    private val deviceDiscoveryManager: DeviceDiscoveryManager by lazy {
        DeviceDiscoveryManager(upnpService, upnpRegistryListener)
    }
    private val volumeManager: VolumeManager by lazy {
        VolumeManager(upnpService)
    }

    // --- Plugin lifecycle ---
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        MediaCastDlnaApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        MediaCastDlnaApi.setUp(binding.binaryMessenger, null)
        if (upnpService != null && context != null) {
            context?.unbindService(serviceConnection)
        }
        upnpService = null
        context = null
    }

    override fun initializeUpnpService(callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                context?.let {
                    val intent = Intent(it, AndroidUpnpServiceImpl::class.java)
                    val bindResult =
                        it.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
                    callback(Result.success(Unit))
                } ?: run {
                    val error =
                        IllegalStateException("Context is not available. Ensure the plugin is attached to the engine.")
                    callback(Result.failure(error))
                }
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun isUpnpServiceInitialized(callback: (Result<Boolean>) -> Unit) {
        pluginScope.launch {
            try {
                val isInitialized = upnpService != null && isServiceBound
                callback(Result.success(isInitialized))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun shutdownUpnpService(callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                if (upnpService != null && context != null && isServiceBound) {
                    try {
                        context!!.unbindService(serviceConnection)
                    } catch (e: IllegalArgumentException) {
                        // Service unbind failed, ignore
                    }
                    isServiceBound = false
                }
                upnpService = null
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    // --- Discovery ---
    override fun startDiscovery(options: DiscoveryOptions, callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                if (!isServiceBound) {
                    throw IllegalStateException("UPnP service is not bound. Please ensure the service is started before calling startDiscovery.")
                }
                deviceDiscoveryManager.startDiscovery()
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun stopDiscovery(callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                deviceDiscoveryManager.stopDiscovery()
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getDiscoveredDevices(callback: (Result<List<DlnaDevice>>) -> Unit) {
        pluginScope.launch {
            try {
                val devices = deviceDiscoveryManager.getDiscoveredDevices()
                callback(Result.success(devices))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun refreshDevice(deviceUdn: DeviceUdn, callback: (Result<DlnaDevice?>) -> Unit) {
        pluginScope.launch {
            try {
                val device = deviceDiscoveryManager.refreshDevice(deviceUdn.value)
                callback(Result.success(device))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getDeviceServices(
        deviceUdn: DeviceUdn, callback: (Result<List<DlnaService>>) -> Unit
    ) {
        pluginScope.launch {
            try {
                val services = deviceDiscoveryManager.getDeviceServices(deviceUdn.value)
                callback(Result.success(services))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun hasService(
        deviceUdn: DeviceUdn, serviceType: String, callback: (Result<Boolean>) -> Unit
    ) {
        pluginScope.launch {
            try {
                val hasService = deviceDiscoveryManager.hasService(deviceUdn.value, serviceType)
                callback(Result.success(hasService))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun isDeviceOnline(deviceUdn: DeviceUdn, callback: (Result<Boolean>) -> Unit) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.success(false))
            return
        }

        pluginScope.launch {
            try {
                val isOnline = withTimeout(5000L) { // 5 second timeout for connectivity check
                    deviceDiscoveryManager.isDeviceOnline(deviceUdn.value)
                }
                callback(Result.success(isOnline))
            } catch (e: TimeoutCancellationException) {
                callback(Result.success(false))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    // --- Media Control (delegating to managers with coroutines) ---
    override fun setMediaUri(
        deviceUdn: DeviceUdn, uri: Url, metadata: MediaMetadata, callback: (Result<Unit>) -> Unit
    ) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.failure(Exception("UPnP service not initialized")))
            return
        }

        pluginScope.launch {
            try {
                val finalMetadata = didlMetadataConverter.toDidlLite(metadata, uri.value)

                withTimeout(30000L) {
                    mediaControlManager.setMediaUri(deviceUdn.value, uri.value, finalMetadata)
                }
                callback(Result.success(Unit))
            } catch (e: TimeoutCancellationException) {
                callback(Result.failure(e))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun setMediaUriWithSubtitles(
        deviceUdn: DeviceUdn,
        uri: Url,
        metadata: MediaMetadata,
        subtitleTracks: List<SubtitleTrack>,
        callback: (Result<Unit>) -> Unit
    ) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.failure(Exception("UPnP service not initialized")))
            return
        }

        pluginScope.launch {
            try {
                val mediaControlManager = MediaControlManager(upnpService)
                val metadataConverter = DefaultDidlMetadataConverter()
                val finalMetadata = metadataConverter.toDidlLite(metadata, uri.value)

                withTimeout(30000) { // 30-second timeout
                    mediaControlManager.setMediaUriWithSubtitlesEnhanced(
                        deviceUdn.value, uri.value, finalMetadata, subtitleTracks
                    )
                }

                callback(Result.success(Unit))
            } catch (e: TimeoutCancellationException) {
                callback(Result.failure(Exception("Operation timed out: ${e.message}")))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun supportsSubtitleControl(
        deviceUdn: DeviceUdn, callback: (Result<Boolean>) -> Unit
    ) {
        pluginScope.launch {
            try {
                val supportsSubtitles =
                    mediaControlManager.checkDeviceSubtitleSupport(deviceUdn.value)
                callback(Result.success(supportsSubtitles))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun setSubtitleTrack(
        deviceUdn: DeviceUdn, subtitleTrackId: String?, callback: (Result<Unit>) -> Unit
    ) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.failure(Exception("UPnP service not initialized")))
            return
        }

        pluginScope.launch {
            try {
                val mediaControlManager = MediaControlManager(upnpService)

                withTimeout(10000) { // 10-second timeout
                    mediaControlManager.setSubtitleTrack(deviceUdn.value, subtitleTrackId)
                }

                callback(Result.success(Unit))
            } catch (e: UnsupportedOperationException) {
                callback(Result.failure(e))
            } catch (e: TimeoutCancellationException) {
                callback(Result.failure(Exception("Operation timed out: ${e.message}")))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getAvailableSubtitleTracks(
        deviceUdn: DeviceUdn, callback: (Result<List<SubtitleTrack>>) -> Unit
    ) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.success(emptyList()))
            return
        }

        pluginScope.launch {
            try {
                val tracks = withTimeout(10000) {
                    mediaControlManager.getAvailableSubtitleTracks(deviceUdn.value)
                }
                callback(Result.success(tracks))
            } catch (e: TimeoutCancellationException) {
                callback(Result.success(emptyList()))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getCurrentSubtitleTrack(
        deviceUdn: DeviceUdn, callback: (Result<SubtitleTrack?>) -> Unit
    ) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.success(null))
            return
        }

        pluginScope.launch {
            try {
                val track = withTimeout(10000) {
                    mediaControlManager.getCurrentSubtitleTrack(deviceUdn.value)
                }
                callback(Result.success(track))
            } catch (e: TimeoutCancellationException) {
                callback(Result.success(null))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun play(deviceUdn: DeviceUdn, callback: (Result<Unit>) -> Unit) {
        if (!isServiceBound || upnpService == null) {
            callback(Result.failure(Exception("UPnP service not initialized")))
            return
        }

        pluginScope.launch {
            try {
                withTimeout(30000L) { // 30 second timeout
                    mediaControlManager.play(deviceUdn.value)
                }
                callback(Result.success(Unit))
            } catch (e: TimeoutCancellationException) {
                callback(Result.failure(e))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun pause(deviceUdn: DeviceUdn, callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                mediaControlManager.pause(deviceUdn.value)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun stop(deviceUdn: DeviceUdn, callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                mediaControlManager.stop(deviceUdn.value)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun seek(
        deviceUdn: DeviceUdn, position: TimePosition, callback: (Result<Unit>) -> Unit
    ) {
        pluginScope.launch {
            try {
                mediaControlManager.seek(deviceUdn.value, position.seconds.toLong())
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun setVolume(deviceUdn: DeviceUdn, volumeLevel: VolumeLevel, callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                volumeManager.setVolume(deviceUdn.value, volumeLevel.percentage)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getVolumeInfo(deviceUdn: DeviceUdn, callback: (Result<VolumeInfo>) -> Unit) {
        pluginScope.launch {
            try {
                val volumeInfo = volumeManager.getVolumeInfo(deviceUdn)
                callback(Result.success(volumeInfo))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun setMute(deviceUdn: DeviceUdn, muteOperation: MuteOperation, callback: (Result<Unit>) -> Unit) {
        pluginScope.launch {
            try {
                volumeManager.setMute(deviceUdn.value, muteOperation.shouldMute)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getPlaybackInfo(deviceUdn: DeviceUdn, callback: (Result<PlaybackInfo>) -> Unit) {
        pluginScope.launch {
            try {
                val playbackInfo = mediaControlManager.getPlaybackInfo(deviceUdn.value)
                callback(Result.success(playbackInfo))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getCurrentPosition(deviceUdn: DeviceUdn, callback: (Result<TimePosition>) -> Unit) {
        pluginScope.launch {
            try {
                val position = mediaControlManager.getCurrentPosition(deviceUdn.value)
                callback(Result.success(TimePosition(position)))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun getTransportState(
        deviceUdn: DeviceUdn, callback: (Result<TransportState>) -> Unit
    ) {
        pluginScope.launch {
            try {
                val transportState = mediaControlManager.getTransportState(deviceUdn.value)
                callback(Result.success(transportState))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
}

File: ios/Classes/MediaCastDlnaPlugin.swift
import Flutter
import UIKit
import Foundation
import Network
import os.log
import UPnAtom

#if canImport(Darwin)
import Darwin
#endif

#if canImport(SystemConfiguration)
import SystemConfiguration
#endif

public class MediaCastDlnaPlugin: NSObject, FlutterPlugin, MediaCastDlnaApi {
    
    // MARK: - Properties
    private var upnpDiscovery: UPnAtomDeviceDiscovery?
    private var discoveredDevices: [String: DlnaDevice] = [:]
    private var isServiceInitialized = false
    private var discoveryTimer: Timer?
    private let logger = OSLog(subsystem: "media_cast_dlna", category: "iOS")
    
    // Managers for different functionality (using UPnAtom)
    private lazy var deviceDiscoveryManager = DeviceDiscoveryManager()
    private lazy var mediaControlManager: UPnAtomMediaControlManager = {
        return UPnAtomMediaControlManager(discovery: upnpDiscovery!)
    }()
    private lazy var volumeManager = VolumeManager()
    private lazy var metadataConverter = DidlMetadataConverter()
    
    // MARK: - Flutter Plugin Registration
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = MediaCastDlnaPlugin()
        MediaCastDlnaApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }
    
    // MARK: - Service Lifecycle
    public func initializeUpnpService() throws {
        os_log("Initializing UPnP service on iOS", log: logger, type: .info)
        
        guard !isServiceInitialized else {
            os_log("UPnP service already initialized", log: logger, type: .info)
            return
        }
        
        // Initialize UPnP discovery using UPnAtom
        upnpDiscovery = UPnAtomDeviceDiscovery()
        upnpDiscovery?.delegate = self
        
        // Initialize managers
        deviceDiscoveryManager.initialize()
        mediaControlManager.initialize()
        volumeManager.initialize()
        
        isServiceInitialized = true
        os_log("UPnP service initialized successfully", log: logger, type: .info)
    }
    
    public func isUpnpServiceInitialized() throws -> Bool {
        return isServiceInitialized
    }
    
    public func shutdownUpnpService() throws {
        os_log("Shutting down UPnP service", log: logger, type: .info)
        
        stopDiscovery()
        upnpDiscovery = nil
        discoveredDevices.removeAll()
        isServiceInitialized = false
        
        os_log("UPnP service shutdown complete", log: logger, type: .info)
    }
    
    // MARK: - Device Discovery
    // MARK: - Device Discovery
    public func startDiscovery(options: DiscoveryOptions) throws {
        guard isServiceInitialized else {
            throw PigeonError(code: "SERVICE_NOT_INITIALIZED", 
                            message: "UPnP service is not initialized", 
                            details: nil)
        }
        
        os_log("Starting device discovery", log: logger, type: .info)
        
        // Start SSDP discovery
        upnpDiscovery?.startDiscovery()
        
        // Set up periodic refresh using timeout as refresh interval
        discoveryTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(options.timeout), repeats: true) { _ in
            self.refreshDeviceList()
        }
        
        os_log("Device discovery started", log: logger, type: .info)
    }
    
    public func stopDiscovery() throws {
        os_log("Stopping device discovery", log: logger, type: .info)
        
        upnpDiscovery?.stopDiscovery()
        discoveryTimer?.invalidate()
        discoveryTimer = nil
        
        os_log("Device discovery stopped", log: logger, type: .info)
    }
    
    public func getDiscoveredDevices() throws -> [DlnaDevice] {
        return Array(discoveredDevices.values)
    }
    
    public func refreshDevice(deviceUdn: String) throws -> DlnaDevice? {
        return deviceDiscoveryManager.refreshDevice(deviceUdn: deviceUdn)
    }
    
    public func getDeviceServices(deviceUdn: String) throws -> [DlnaService] {
        return deviceDiscoveryManager.getDeviceServices(deviceUdn: deviceUdn)
    }
    
    public func hasService(deviceUdn: String, serviceType: String) throws -> Bool {
        return deviceDiscoveryManager.hasService(deviceUdn: deviceUdn, serviceType: serviceType)
    }
    
    public func browseContentDirectory(deviceUdn: String, parentId: String, startIndex: Int64, requestCount: Int64) throws -> [MediaItem] {
        return deviceDiscoveryManager.browseContentDirectory(
            deviceUdn: deviceUdn,
            parentId: parentId,
            startIndex: startIndex,
            requestCount: requestCount
        )
    }
    
    public func searchContentDirectory(deviceUdn: String, containerId: String, searchCriteria: String, startIndex: Int64, requestCount: Int64) throws -> [MediaItem] {
        return deviceDiscoveryManager.searchContentDirectory(
            deviceUdn: deviceUdn,
            containerId: containerId,
            searchCriteria: searchCriteria,
            startIndex: startIndex,
            requestCount: requestCount
        )
    }
    
    // MARK: - Media Control
    public func setMediaUri(deviceUdn: String, uri: String, metadata: MediaMetadata, completion: @escaping (Result<Void, Error>) -> Void) {
        os_log("Setting media URI for device %@: %@", log: logger, type: .info, deviceUdn, uri)
        
        guard isServiceInitialized else {
            completion(.failure(PigeonError(code: "SERVICE_NOT_INITIALIZED", 
                                          message: "UPnP service is not initialized", 
                                          details: nil)))
            return
        }
        
        Task {
            do {
                let didlMetadata = try metadataConverter.toDidlLite(metadata: metadata, uri: uri)
                try await mediaControlManager.setMediaUri(deviceUdn: deviceUdn, uri: uri, metadata: didlMetadata)
                completion(.success(()))
            } catch {
                os_log("Failed to set media URI: %@", log: logger, type: .error, error.localizedDescription)
                completion(.failure(error))
            }
        }
    }
    
    public func setMediaUriWithSubtitles(deviceUdn: String, uri: String, metadata: MediaMetadata, subtitleTracks: [SubtitleTrack], completion: @escaping (Result<Void, Error>) -> Void) {
        os_log("Setting media URI with subtitles for device %@", log: logger, type: .info, deviceUdn)
        
        guard isServiceInitialized else {
            completion(.failure(PigeonError(code: "SERVICE_NOT_INITIALIZED", 
                                          message: "UPnP service is not initialized", 
                                          details: nil)))
            return
        }
        
        Task {
            do {
                let didlMetadata = try metadataConverter.toDidlLite(metadata: metadata, uri: uri)
                try await mediaControlManager.setMediaUriWithSubtitles(
                    deviceUdn: deviceUdn,
                    uri: uri,
                    metadata: didlMetadata,
                    subtitleTracks: subtitleTracks
                )
                completion(.success(()))
            } catch {
                os_log("Failed to set media URI with subtitles: %@", log: logger, type: .error, error.localizedDescription)
                completion(.failure(error))
            }
        }
    }
    
    public func supportsSubtitleControl(deviceUdn: String) throws -> Bool {
        return mediaControlManager.supportsSubtitleControl(deviceUdn: deviceUdn)
    }
    
    public func setSubtitleTrack(deviceUdn: String, subtitleTrackId: String?, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.setSubtitleTrack(deviceUdn: deviceUdn, subtitleTrackId: subtitleTrackId)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func getAvailableSubtitleTracks(deviceUdn: String) throws -> [SubtitleTrack] {
        return mediaControlManager.getAvailableSubtitleTracks(deviceUdn: deviceUdn)
    }
    
    public func getCurrentSubtitleTrack(deviceUdn: String) throws -> SubtitleTrack? {
        return mediaControlManager.getCurrentSubtitleTrack(deviceUdn: deviceUdn)
    }
    
    public func play(deviceUdn: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.play(deviceUdn: deviceUdn)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func pause(deviceUdn: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.pause(deviceUdn: deviceUdn)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func stop(deviceUdn: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.stop(deviceUdn: deviceUdn)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func seek(deviceUdn: String, positionSeconds: Int64, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.seek(deviceUdn: deviceUdn, positionSeconds: positionSeconds)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func next(deviceUdn: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.next(deviceUdn: deviceUdn)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func previous(deviceUdn: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await mediaControlManager.previous(deviceUdn: deviceUdn)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Volume Control
    public func setVolume(deviceUdn: String, volume: Int64, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await volumeManager.setVolume(deviceUdn: deviceUdn, volume: volume)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    public func getVolumeInfo(deviceUdn: String) throws -> VolumeInfo {
        return volumeManager.getVolumeInfo(deviceUdn: deviceUdn)
    }
    
    public func setMute(deviceUdn: String, muted: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                try await volumeManager.setMute(deviceUdn: deviceUdn, muted: muted)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Playback Info
    public func getPlaybackInfo(deviceUdn: String) throws -> PlaybackInfo {
        return mediaControlManager.getPlaybackInfo(deviceUdn: deviceUdn)
    }
    
    public func getCurrentPosition(deviceUdn: String) throws -> Int64 {
        return mediaControlManager.getCurrentPosition(deviceUdn: deviceUdn)
    }
    
    public func getTransportState(deviceUdn: String) throws -> TransportState {
        return mediaControlManager.getTransportState(deviceUdn: deviceUdn)
    }
    
    // MARK: - Platform Info
    public func getPlatformVersion() throws -> String {
        return "iOS " + UIDevice.current.systemVersion
    }
    
    public func isUpnpAvailable() throws -> Bool {
        return true // UPnP is available on iOS
    }
    
    public func getNetworkInterfaces() throws -> [String] {
        var interfaces: [String] = []
        
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0 else { return interfaces }
        guard let firstAddr = ifaddr else { return interfaces }
        
        for ifptr in sequence(first: firstAddr, next: { $0.pointee.ifa_next }) {
            let interface = ifptr.pointee
            let addrFamily = interface.ifa_addr.pointee.sa_family
            
            if addrFamily == UInt8(AF_INET) || addrFamily == UInt8(AF_INET6) {
                let name = String(cString: interface.ifa_name)
                if !interfaces.contains(name) {
                    interfaces.append(name)
                }
            }
        }
        
        freeifaddrs(ifaddr)
        return interfaces
    }
    
    // MARK: - Private Methods
    private func refreshDeviceList() {
        upnpDiscovery?.refreshDevices()
    }
}

// MARK: - UPnP Discovery Delegate
extension MediaCastDlnaPlugin: UPnPDeviceDiscoveryDelegate {
    func deviceDiscovered(_ device: DlnaDevice) {
        discoveredDevices[device.udn] = device
        deviceDiscoveryManager.addDevice(device)
        os_log("Device discovered: %@", log: logger, type: .info, device.friendlyName)
    }
    
    func deviceRemoved(_ deviceUdn: String) {
        discoveredDevices.removeValue(forKey: deviceUdn)
        deviceDiscoveryManager.removeDevice(udn: deviceUdn)
        os_log("Device removed: %@", log: logger, type: .info, deviceUdn)
    }
}
